### PR TITLE
✨ : Animation controller now has ability to repeat animation 'n' no. of times.

### LINKED ISSUE #53262: AnimationController should have the ability to repeat n times
## Use case

AnimationController should provide a simple way to repeat n times, currently, this is not possible from what we could find.
This is the nearest and simplest solution: https://stackoverflow.com/questions/59462290/repeating-animations-specific-times-e-g-20-times-by-flutter/59463013#59463013

If an animation should repeat n times, it should do the animation and then jump back to `0` (or `1` for reverse) and do the animation again. Until it reached `1` (or `0` for reverse) n times and then stop there and send a `completion` event in the status listener.

## Proposal

```dart
controller = AnimationController(
  vsync: this,
  duration: const Duration(seconds: 2),
)
..repeat(times: 3)
..addStatusListener((status) {
    if (status == AnimationStatus.completed) {
      print("Has repeated 3 times. Execution took 6 seconds.");
    }
});
```

### PR DESCRIPTION
**Pull request by aliasgar4558**
Right now animation controller does not have any way by which user can repeat animation with specific no. of times. Adding the changes in existing `repeat` method in which users can specific no. of times they want to repeat the animation. If not specified, it would simulate infinitely. [Existing repeat behaviour]

This PR fixes : https://github.com/flutter/flutter/issues/53262

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

### CODE REVIEW

File: packages/flutter/lib/src/animation/animation_controller.dart
Comment (nate-thegrate) [inline packages/flutter/lib/src/animation/animation_controller.dart:708]:
```suggestion
  /// If a value is passed to [repeatCount], the animation will perform that many
  /// iterations before stopping. Otherwise, the animation repeats indefinitely.
```

File: packages/flutter/lib/src/animation/animation_controller.dart
Comment (nate-thegrate) [inline packages/flutter/lib/src/animation/animation_controller.dart:711]: We should probably update this documentation, now that there's a way for the `TickerFuture` to complete.

```suggestion
  /// Returns a [TickerFuture] that never completes, unless a [repeatCount] is specified.
  /// The [TickerFuture.orCancel] future completes with an error when the animation is
  /// stopped (e.g. with [stop]).
```

File: packages/flutter/lib/src/animation/animation_controller.dart
Comment (nate-thegrate) [inline packages/flutter/lib/src/animation/animation_controller.dart:977]:
```suggestion
    this.directionSetter,
    this.repeatCount,
  )  : assert(
```

Small nit: since this is a private class that never omits the final argument, the brackets aren't needed here.

File: packages/flutter/test/animation/animation_controller_test.dart
Comment (nate-thegrate) [inline packages/flutter/test/animation/animation_controller_test.dart:1248]: Thanks for adding these tests, they look great!

File: packages/flutter/lib/src/animation/animation_controller.dart
Comment (nate-thegrate) [inline packages/flutter/lib/src/animation/animation_controller.dart:1036]: Instead of computing these during each function call, could we set them as `late final` class members or something similar?

Also, is there a reason why we're rounding to 3 decimal places here?

File: packages/flutter/lib/src/animation/animation_controller.dart
Comment (nate-thegrate) [inline packages/flutter/lib/src/animation/animation_controller.dart:741]:
```suggestion
    return _startSimulation(_RepeatingSimulation(_value, min, max, reverse, period!, _directionSetter, repeatTimes));
```

(nit: only use a trailing comma when arguments are on separate lines)

**Review by nate-thegrate — COMMENTED** [review]
Thanks for contributing; overall I think this is a fantastic addition to the `.repeat()` API!

I decided to give a lot of feedback, so please take a look when you have a chance :)

File: packages/flutter/lib/src/animation/animation_controller.dart
Comment (aliasgar4558) [reply packages/flutter/lib/src/animation/animation_controller.dart:1038]: Added private variable to calculate exit time in seconds only once for entire simulation.

Reason behind rounding it to 3 decimal points is, usually we don't need 7-8 decimal points value as its already in seconds so 3 decimal points shall be enough. Also, depending upon the device performance and all - simulation might differ if we compare whole decimal value. Please suggest your thoughts.

File: packages/flutter/lib/src/animation/animation_controller.dart
Comment (aliasgar4558) [reply packages/flutter/lib/src/animation/animation_controller.dart:723]: okay sure, updated.

**Comment by aliasgar4558** [comment]
@nate-thegrate - Thanks for the review. I am glad you found this changes a good add to the framework. I will update your review comments in some time today.

File: packages/flutter/lib/src/animation/animation_controller.dart
Comment (nate-thegrate) [inline packages/flutter/lib/src/animation/animation_controller.dart:1011]:
```suggestion
  /// If [reverse] is true, the duration of a full cycle is doubled.
  late final _effectiveRepeatCount = repeatCount! * (reverse ? 2 : 1);

  late final double _exitTimeInSeconds = repeatCount! * _periodInSeconds - _initialT;
```

File: packages/flutter/lib/src/animation/animation_controller.dart
Comment (nate-thegrate) [inline packages/flutter/lib/src/animation/animation_controller.dart:1047]:
```suggestion
    return repeatCount == null || timeInSeconds >= _exitTimeInSeconds;
```

**Review by nate-thegrate — COMMENTED** [review]
Thanks for making those updates!

I don't think rounding is a good idea here, since it doesn't make things any more simple under the hood:

| base ten | binary                             |
|----------|------------------------------------|
| 5        | 101                                |
| 256      | 10000000                           |
| 0.125    | 0.001                              |
| 0.1      | 0.0001100110011001100110011001100… |

(See [this thread](https://github.com/flutter/flutter/pull/150764#discussion_r1658638669) for a few more details.)

<br>

Once we streamline the logic here, I think we'll be good to go!

File: packages/flutter/lib/src/animation/animation_controller.dart
Comment (nate-thegrate) [reply packages/flutter/lib/src/animation/animation_controller.dart:1047]: I just noticed some mistakes in the suggestions I made!

Glad I was able to catch those before you pushed an update 😅

Comment (aliasgar4558) [reply packages/flutter/lib/src/animation/animation_controller.dart:1047]: Any reason to remove `_initialT` from calculation?

File: packages/flutter/lib/src/animation/animation_controller.dart
Comment (aliasgar4558) [reply packages/flutter/lib/src/animation/animation_controller.dart:1011]: Make sense, updated.

File: packages/flutter/lib/src/animation/animation_controller.dart
Comment (aliasgar4558) [reply packages/flutter/lib/src/animation/animation_controller.dart:1047]: hehe, No problem. Thanks for guiding to the better solution.

File: packages/flutter/lib/src/animation/animation_controller.dart
Comment (nate-thegrate) [inline packages/flutter/lib/src/animation/animation_controller.dart:1003]:
```suggestion
  late final double _exitTimeInSeconds = (_effectiveRepeatCount * _periodInSeconds) - _initialT;
```

Small nit: I think this should be one line. There's a similar example from the [style guide](https://github.com/flutter/flutter/blob/38af6be9fb29bb87b9b2fb7853fe22b4287ea917/docs/contributing/Style-guide-for-Flutter-repo.md#prefer-a-maximum-line-length-of-80-characters):

```dart
// BAD (breaks after assignment operator)
final List<FooBarBaz> _members =
  <FooBarBaz>[const Quux(), const Qaax(), const Qeex()];

// BETTER (only slightly goes over 80 chars)
final List<FooBarBaz> _members = <FooBarBaz>[const Quux(), const Qaax(), const Qeex()];
```

File: packages/flutter/lib/src/animation/animation_controller.dart
Comment (nate-thegrate) [inline packages/flutter/lib/src/animation/animation_controller.dart:1030]: Thanks very much for cleaning this up 😊

**Review by nate-thegrate — APPROVED** [review]
One small suggestion; otherwise LGTM 👍

Thanks for putting in the work here!

File: packages/flutter/lib/src/animation/animation_controller.dart
Comment (aliasgar4558) [reply packages/flutter/lib/src/animation/animation_controller.dart:1030]: Welcome. Also updated comment.

File: packages/flutter/lib/src/animation/animation_controller.dart
Comment (aliasgar4558) [reply packages/flutter/lib/src/animation/animation_controller.dart]: hmm thanks for pointing out the style guidelines. I have updated as per suggestion.

**Comment by nate-thegrate** [comment]
Hi @aliasgar4558, thought I'd give a quick heads-up:

This PR will be merged once we get a second review. No guarantees as to when that will happen, but it'll probably be within a couple of days!

<br>

I also see several merge commits here. According to the [Tree Hygeine docs](https://github.com/flutter/flutter/blob/348fdeebb31496f6bb10c576d6d973748b422a8f/docs/contributing/Tree-hygiene.md#using-git), it's better to do an "update with rebase" instead.
(And unless tests are flaking, it might not be worth it to kick off all the checks again.)

**Comment by aliasgar4558** [comment]
@nate-thegrate - Okay sure, No worries. 

Moreover, I didn't ger you when you say update with rebase. Are you suggesting that we should not update this branch with parent very oftenly now? And when we initiate second review, we can do that?

**Comment by nate-thegrate** [comment]
Yeah, generally, repeated merge commits might not be extremely helpful, and according to the wiki, things are more likely to go well when you do an "update with rebase".

update with rebase

File: packages/flutter/lib/src/animation/animation_controller.dart
Comment (nate-thegrate) [inline packages/flutter/lib/src/animation/animation_controller.dart:725]: flutter_animate just uses the word `count`, instead of `repeatCount`.

```suggestion
    int? count,
```

File: packages/flutter/lib/src/animation/animation_controller.dart
Comment (nate-thegrate) [inline packages/flutter/lib/src/animation/animation_controller.dart:1003]: If `reverse` is true, we can choose to _not_ double the duration: that way, we can pass an odd number to `count` in order for it to end in the "completed" state.

```suggestion
  late final double _exitTimeInSeconds = (repeatCount! * _periodInSeconds) - _initialT;
```

File: packages/flutter/test/animation/animation_controller_test.dart
Comment (nate-thegrate) [inline packages/flutter/test/animation/animation_controller_test.dart:1313]:
```suggestion
        controller.repeat(reverse: true, repeatCount: 4);
```

**Review by nate-thegrate — CHANGES_REQUESTED** [review]
Hey Aliasgar, sorry to pull a switcheroo after sending an approval, but I've come across some important information:

There's a popular independent package on pub.dev called [flutter_animate](https://pub.dev/packages/flutter_animate) that already implemented an AnimationController extension method with a "repeat count" option.

I like the way they chose to implement it, and I think we should tweak this PR to match for ease of migration.

File: packages/flutter/lib/src/animation/animation_controller.dart
Comment (aliasgar4558) [reply packages/flutter/lib/src/animation/animation_controller.dart:1003]: but ideally, people would think when `count` given as 2, the animation shall repeat twice. If we remove this doubling logic, then if user passed 2 then it will only repeat once, isn't it? 

I think we shall keep this doubling logic at controller level to make developer experience more streamline. Please suggest your thoughts on this one.,

File: packages/flutter/test/animation/animation_controller_test.dart
Comment (aliasgar4558) [reply packages/flutter/test/animation/animation_controller_test.dart:1313]: This also serves the same concern. If I pass 4 as count, I would intend this simulation to repeat 4 times as I am using repeat method.

File: packages/flutter/lib/src/animation/animation_controller.dart
Comment (nate-thegrate) [reply packages/flutter/lib/src/animation/animation_controller.dart:1003]: I think it would be nice to allow the following:

```
forward, reverse, forward, reverse, forward (and then stop)
```

Removing the doubling logic allows the animation to terminate in the "completed" state, and it matches the current behavior shown in a [flutter_animate example](https://youtu.be/lxkFL9disq0?t=544).

Comment (aliasgar4558) [reply packages/flutter/lib/src/animation/animation_controller.dart:1003]: hmm alright, if you say so then I will update the changes accordingly. Please allow me some time

**Comment by aliasgar4558** [comment]
@nate-thegrate Updated code, documentation & tests as per review comments.

**Review by nate-thegrate — APPROVED** [review]
Looks fantastic, thank you so much!

**Comment by aliasgar4558** [comment]
Thanks @nate-thegrate for the valuable insights & suggestions.

**Comment by nate-thegrate** [comment]
There's a decent chance that rebasing with the latest updates to master could fix the failing check.

<details> <summary>(Pushing an update is great as a workaround for flaky tests, but FYI I get an email each time it happens 🙃)</summary>

lots of commits

</details>

**Comment by nate-thegrate** [comment]
> So you recommend not to rebase very often?
> And shall I rebase with master branch now to get the latest changes??

Yes sir!

> Btw, it's a long thread of emails sorry

No worries! 🤠👍

**Comment by aliasgar4558** [comment]
@chunhtai - any update on this one??

**Review by chunhtai — APPROVED** [review]
LGTM

**Comment by aliasgar4558** [comment]
@nate-thegrate @chunhtai - thanks 👍🏼

**Pull request merged**
